### DISCUSSION TITLE
Changing the colormap for cameragui

### DISCUSSION THREAD

**Discussion by mario-b-amaro** [Ulm-IQO/qudi-iqo-modules#95] [Q&A] [answered]
How could one change the colormap of camera_gui? The default seems to be 'inferno', however it would be nice if it would be possible to change it. I initially thought it could be changed 'camera_logic.py', in the function draw_3d_image. Sadly, even after changing the cmap there and reseting qudi, the camera_gui still plots the camera image in the default colormap. Does anyone know the way to do it? Thank you!

**Comment by Neverhorst** [comment, marked as answer]
Hi @mario-b-amaro ,

The plotting code you can find in any logic module is just responsible for creating a thumbnail image using `matplotlib` that is saved to file alongside the measurement data. `matplotlib` has a nice interface to set the colorscale via keyword (`inferno`, `rainbow` etc.). But the plotting widgets of the GUI modules do not use `matplotlib` but `Qt for Python (PySide2)`.
Currently there is no concise interface to change the colorscale of a GUI plotwidget (from `qudi.util.widgets.plotting`) via a simple function call. It would be nice to have this functionality, so feel free to [file a feature request](https://github.com/Ulm-IQO/qudi-core/issues/new/choose) in `qudi-core`.

If you really need to change the colorscale now, you can do it the ugly way by altering `qudi.gui.camera.cameragui`; more specifically by appending the following lines to the `CameraMainWindow.__init__`. However I advise against such quick-and-dirty changes to the code because they are non-maintainable and cause your code to diverge from the upstream repository. It's better to fix this in `qudi-core` and make it generally configurable.

```Python
self.image_widget.image_item.setLookupTable(ColorScaleViridis().lut)
grad = QtGui.QLinearGradient(0, 0, 0, 1)
grad.setCoordinateMode(QtGui.QGradient.ObjectMode)
for stop, color in zip(*ColorScaleViridis().colormap.getStops('byte')):
    grad.setColorAt(stop, QtGui.QColor(*color))
self.image_widget.colorbar_widget.colorbar.setOpts(brush=QtGui.QBrush(grad))
```

`ColorScaleViridis` is just an example of a colorscale imported from `qudi.util.colordefs`. If the colorscale you want can not be found in this module, you can create your own by subclassing `qudi.util.colordefs.ColorScale` and providing the `COLORS` class attribute. Have a look at [the module in `qudi-core`](https://github.com/Ulm-IQO/qudi-core/blob/main/src/qudi/util/colordefs.py) to see how this is done.
